pkg/ctl/create/cluster.go

```diff
@@ -424,6 +417,75 @@ func doCreateCluster(cmd *cmdutils.Cmd, ngFilter *filter.NodeGroupFilter, params
 	return nil
 }
 
+// installKarpenter prepares the environment for Karpenter, by creating the following resources:
+// - iam roles and profiles
+// - service account
+// - identity mapping
+// then proceeds with installing Karpenter using Helm.
+func installKarpenter(ctl *eks.ClusterProvider, cfg *api.ClusterConfig, stackManager manager.StackManager, meta *api.ClusterMeta, clientSet *kubeclient.Clientset) error {
+	if cfg.Karpenter == nil {
+		return nil
+	}
+	clientSetGetter := &kubernetes.CallbackClientSet{
+		Callback: func() (kubernetes.Interface, error) {
+			return ctl.NewStdClientSet(cfg)
+		},
+	}
+
+	// create the needed service account before Karpenter, otherwise, Karpenter will fail to be created.
+	parsedARN, err := arn.Parse(cfg.Status.ARN)
+	if err != nil {
+		return fmt.Errorf("unexpected invalid ARN: %q, %w", cfg.Status.ARN, err)
+	}
+	roleArn := fmt.Sprintf("arn:aws:iam::%s:policy/%s-%s", parsedARN.AccountID, builder.KarpenterManagedPolicy, cfg.Metadata.Name)
+	karpenterServiceAccountTaskTree := stackManager.NewTasksToCreateIAMServiceAccounts([]*api.ClusterIAMServiceAccount{
+		{
+			ClusterIAMMeta: api.ClusterIAMMeta{
+				Name:      karpenter.DefaultServiceAccountName,
+				Namespace: karpenter.DefaultNamespace,
+			},
+			AttachRoleARN: roleArn,
+		},
+	}, nil, clientSetGetter)
+	logger.Info(karpenterServiceAccountTaskTree.Describe())
+	if errs := karpenterServiceAccountTaskTree.DoAllSync(); len(errs) > 0 {
+		logger.Warning("%d error(s) occurred while creating service account for Karpenter, you may wish to check your Cluster for further information", len(errs))
+		for _, err := range errs {
+			ufe := &api.UnsupportedFeatureError{}
+			if errors.As(err, &ufe) {
+				logger.Critical(ufe.Message)
+			}
+			logger.Critical("%s\n", err.Error())
+		}
+		return fmt.Errorf("failed to create Karpenter service account %q", meta.Name)
+	}
+
+	// create identity mapping for EC2 nodes to be able to join the cluster.
+	acm, err := authconfigmap.NewFromClientSet(clientSet)
+	if err != nil {
+		return fmt.Errorf("failed to create client for auth config: %w", err)
+	}
+	identityArn := fmt.Sprintf("arn:aws:iam::%s:role/%s-%s", parsedARN.AccountID, builder.KarpenterNodeRoleName, cfg.Metadata.Name)
+	id, err := iam.NewIdentity(identityArn, authconfigmap.RoleNodeGroupUsername, authconfigmap.RoleNodeGroupGroups)
+	if err != nil {
+		return fmt.Errorf("failed to create new identity: %w", err)
+	}
+	if err := acm.AddIdentity(id); err != nil {
+		return fmt.Errorf("failed to add new identity: %w", err)
+	}
+
+	// install karpenter onto the cluster.
+	installer, err := karpenteractions.NewInstaller(cfg, ctl, stackManager)
+	if err != nil {
+		return fmt.Errorf("failed to create installer: %w", err)
+	}
+	if err := installer.Create(); err != nil {
+		return fmt.Errorf("failed to install Karpenter: %w", err)
+	}
+
+	return nil
+}
+
```

Comment:
can some of this just not live in `ctl/create/`? Might make unit testing easier

Reply:
Pfff, maybe. There is a lot of context which would have to be passed around. But I can try.

Reply:
yay, this turned out nicely. Also, kudos to @nikimanoledaki for first flagging it as an issue and booo to me for not extracting it. :D Thank you both. :)